rov/movement/mapper/Complex_1.py

```diff
@@ -0,0 +1,275 @@
+        # the pseudo inverse of self.matrix used to find the least square solution
+        self.pseudo_inverse_matrix = None
+        # list of disabled thrusters used to determine when matrix and pseudo_inverse_matrix need to be updated
+        self.disabled = [0, 0, 0, 0, 0, 0, 0, 0]
+        # The last thrust map returned by the calculate function
+        self.map = None
+
+        self.generate_matrix()
+
+    def calculate(self, desired_thrust, disabled_thrusters=None, disable_limiting=False):
+        """
+        Calculate the needed thrust for each thruster to achieve desired
+        :param desired_thrust: The 6 dimensional vector which we want to achieve vector as 6x1 matrix
+        :param disabled_thrusters: list of 8 items each corresponding to a thruster (0 meaning enabled, 1 disabled)
+        :return: the thrust map generated by solving for the least square solution of the equation
+        """
+        # In the case of a thruster malfunction this allows for the pseudo_inverse_matrix to be recalculated
+        # to account for the thruster that no longer works
+        if disabled_thrusters != self.disabled:
+            self.disabled = disabled_thrusters
+            self.generate_matrix()
+
+        self.map = self.pseudo_inverse_matrix.dot(desired_thrust)
+
+        self.normalize()
+        initial_power, limitPower = self.calc_thrust_power(self.map)
+        #limit power if necessary: 
+        global FINAL_POWER
+        self.FINAL_POWER = initial_power
+        iteration = 0
+        while limitPower == 1 and disable_limiting == False:
+            if iteration > 3:
+                print('Limit power function iteration limit exceeded, assume values are close enough.')
+                break
+            self.FINAL_POWER = self.limit_power(initial_power)
+            self.FINAL_POWER, limitPower = self.calc_thrust_power(self.map)
+            print('Power was limited, force vector changed!')
+        return self.map
+
+    def generate_matrix(self):
+        """
+        Generate the pseudo-inverse of the matrix to be used in the calculation
+        :return: the pseud-inverse of self.matrix
+        """
+        # Calculate the cross product between the location of each thruster and the direction it points in
+        rot = np.transpose(np.cross(np.transpose(self.ROTATION), np.transpose(self.thruster_layout), 1))
+        self.matrix = np.concatenate((self.ROTATION, rot))
+        for thruster in range(8):
+            if self.disabled[thruster]:
+                self.matrix[:, thruster] = 0.0
+        self.pseudo_inverse_matrix = linalg.pinv(self.matrix)
+        return self.pseudo_inverse_matrix
+
+    def normalize(self):
+        """
+        Normalize the values of the thrust map to be in the range [-1, 1] if necessary
+        :return: None
+        """ 
+        max_val = np.amax(np.abs(self.map))
+        if max_val > 1:
+            self.map /= max_val
+
+    def limit_power(self, initialPower):

```

Comment:
What helps with outside users understanding how to use Python classes is identifying them as 'private' if they are only supposed to be used by the class. Python convention is to begin variables and functions with an underscore ( _ ).